From 748414784f71126b093aa7466908e00f71a7b046 Mon Sep 17 00:00:00 2001
From: Mike Perry <mikeperry-git@torproject.org>
Date: Fri, 27 Mar 2015 12:57:37 -0700
Subject: [PATCH] Bug 15482: Don't abandon circuits that are still in use for
 browsing.

Only applies to connections with SOCKS auth set, so that non-web Tor
activity is not affected.

Simpler version of Nick's patch because the randomness worried me, and I'm not
otherwise sure why we want a max here.
---
 src/or/circuituse.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/src/or/circuituse.c b/src/or/circuituse.c
index d0d31ad..6cce4bf 100644
--- a/src/or/circuituse.c
+++ b/src/or/circuituse.c
@@ -2264,8 +2264,15 @@ connection_ap_handshake_attach_chosen_circuit(entry_connection_t *conn,
 
   base_conn->state = AP_CONN_STATE_CIRCUIT_WAIT;
 
-  if (!circ->base_.timestamp_dirty)
-    circ->base_.timestamp_dirty = time(NULL);
+  if (!circ->base_.timestamp_dirty) {
+    circ->base_.timestamp_dirty = approx_time();
+  } else if ((conn->entry_cfg.isolation_flags & ISO_SOCKSAUTH) &&
+             (conn->socks_request->usernamelen ||
+              conn->socks_request->passwordlen)) {
+    /* When stream isolation is in use and controlled by an application
+     * we are willing to keep using the stream. */
+    circ->base_.timestamp_dirty = approx_time();
+  }
 
   pathbias_count_use_attempt(circ);
 
-- 
1.9.1

